fix(angular1_router): support templateUrl components

petebacondarwin opened the pull request:
Closes #4633

btford commented:
Can you rebase?

btford commented on modules/angular1_router/test/integration/router_spec.js:
```diff
@@ -90,9 +116,11 @@ describe('router', function () {
       controller: getController(options),
-    }
+    };
+    if (options.template) definition.template = options.template;
```
Can we put this in a block?

petebacondarwin replied:
Right-oh!

btford commented on modules/angular1_router/src/ng_outlet.ts:
```diff
@@ -154,13 +154,14 @@ function ngOutletDirective($animate, $q: ng.IQService, $router) {
-        this.controller.$$routeParams = instruction.params;
         this.controller.$$template =
-            '<' + dashCase(componentName) + ' router="$$router"></' + dashCase(componentName) + '>';
+            '<' + dashCase(componentName) + ' $router="$$router"></' + dashCase(componentName) + '>';
```
why change router to `$router` in this binding?

petebacondarwin replied:
I felt that `$router` indicated more clearly that this is an Angular provided thing.

brandonroberts replied:
If someone injects the $router will be overridden by this binding?

petebacondarwin replied:
@brandonroberts - they don't actually affect each other. The scope that is created by the `ngOutlet` directive is separate from the outside scope and from the isolate scope inside the component. Moreover on that scope the value is called `$$router` so there is no collision.

In terms of the binding on the component, the attribute that is injected by the `ngOutlet` directive, which **is** called `$router` can be configured on the binding to be something different on the component if you wish. E.g.

```
bindings: {
  $router: '=localRouter'
}
```

Finally, I have, in any case, added a commit to this PR that actually renames the `$router` service to `$rootRouter`, since that is actually what it is. This is analogous to the `$scope` and `$rootScope` situation with scopes.

brandonroberts replied:
Ok. Thanks for the explanation. I think renaming it to $rootRouter does make it a little more clear.

btford commented:
ping @brandonroberts in case he has insights

petebacondarwin committed:
fix(angular1_router): support templateUrl components

petebacondarwin commented:
I have updated, rebase and split the PR into three commits.
The third commit is new and renames `$router` service to `$rootRouter`.

petebacondarwin committed:
refact(angular1_router): make the `$$router` binding one-time

This binding is never going to change so we can make it a one-time binding

btford commented:
@petebacondarwin can you fix formatting issues? https://travis-ci.org/angular/angular/jobs/108645402#L551

petebacondarwin commented:
@btford rebased and fixed formatting issue

petebacondarwin committed:
fix(angular1_router): rename `$route` service to `$rootRouter`

The singleton service that represents the top level router was called
`$router` but this is confusing since there are actually lots of routers,
which depend upon where you are in the DOM. This is similar to the situation
with scopes.

This commit clarifies this singleton by renaming it to `$rootRouter`.

BREAKING CHANGE:

The `$router` injectable service has been renamed to `$rootRouter`

princemaple commented:
Is this done? Is it the last blocker for angular1 router?

petebacondarwin commented:
We are waiting on it being merged into master.

petebacondarwin committed:
fix(angular1_router): rename `router` component binding to `$router`

The current router is passed to the current component via a binding.
To indicate that this is an angular provided object, this commit
renames the binding to `$router`.

BREAKING CHANGE:

The recently added binding of the current router to the current component
has been renamed from `router` to `$router`.

So now the recommended set up for your bindings in your routed component
is:

```js
{
  ...
  bindings: {
    $router: '<'
  }
}
```

mary-poppins commented:
Merging PR #6978 on behalf of @vsavkin to branch [presubmit-vsavkin-pr-6978](https://github.com/angular/angular/tree/presubmit-vsavkin-pr-6978).

petebacondarwin commented:
Woohoo

char0n commented:
Could somebody please provide built ngComponentRouter in ES5  ? Or direct my to the docs where I can find how to build it. 

Thanks

brandonroberts commented:
@char0n See: https://github.com/angular/angular.js/issues/12926#issuecomment-147137738

alexgoldstone commented:
@char0n Also see https://github.com/angular/angular.js/issues/12926#issuecomment-183674278 for installing via NPM.

char0n commented:
@brandonroberts, @alexgoldstone : thanks. Will try tommorow

DrewynV commented:
__________________________________

Hi! We just wanted to let you know we receievd your email. Thank you for submitting your request. Our aim is to get back to you within 24 hours at the latest. If you have any additional information to add to your request, please reply to this email.


Also please visit our knowledgebase (http://www.support.magicdust.com.au) to browse for further info, we are adding to it all the time.

Or here is our industry report about the 11 Essentials for a Successful Website (http://www.magicdust.com.au/web-design/.

We look forward to helping you succeed online.

Thanks

The Team @ Magicdust Digital

http://support.magicdust.com.au/

Ph (02) 9979 7474
M-F | 9 am-5:00 pm AEST (AEDT in Summer) excluding Public Holidays

------------------------------------------------------------------------------------------------------------
This message was sent to  reply@reply.github.com in reference to Case #166543.
------------------------------------------------------------------------------------------------------------


[[e3bb317eb17828025a6f04b92694776fd28e8af9-673074165]]
